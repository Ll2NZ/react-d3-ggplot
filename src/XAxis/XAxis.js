import React              from "react";
import { Component }      from "react";
import PropTypes          from "prop-types";
import { select }         from "d3-selection";
import { axisBottom }     from "d3-axis";
import { ScalesConsumer } from "../Context/";

class XAxis extends Component{
    static contextType = ScalesConsumer;

    static propTypes = {
        createScaleType: PropTypes.func.isRequired
    }

    findXAxis(){
        // first create scale object
        const { aes }       = this.context;
        const componentName = this.constructor.name;
        const xScale        = this.props.createScaleType(aes[0], componentName);

        // ajust x-axis to bottom
        const { dimensions } = this.context;
        const axisLocation   = `translate(0, ${dimensions.height - dimensions.padding})`;

        // select node returned by component and appends x-axis
        select(this.node)
            .attr("transform", axisLocation)
            .call(axisBottom(xScale));
    }

    render(){
        return(
            <g
                ref={ node => this.node = node }
                className={ this.props.className }
            >
            </g>
        );
    }

    componentDidMount(){
        this.findXAxis();
    }

    componentDidUpdate(){
        this.findXAxis();
    }
}

export default XAxis;
