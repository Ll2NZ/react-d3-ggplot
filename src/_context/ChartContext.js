import { createContext } from "react";

const ChartContext = createContext({
    data: [],
    aes: [],
    dimensions: {}
});

export default ChartContext;
