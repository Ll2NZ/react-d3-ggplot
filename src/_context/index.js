export { default as ChartContext } from "./ChartContext.ts";
