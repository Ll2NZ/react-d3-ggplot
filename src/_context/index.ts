export { default as ChartContext } from "./ChartContext";
export { IAppContext } from "./ChartContext";
