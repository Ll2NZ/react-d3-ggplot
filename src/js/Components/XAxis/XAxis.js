import React           from "react";
import { Component }   from "react";
import PropTypes       from "prop-types";
import { select }      from "d3-selection";
import { axisBottom }  from "d3-axis";
import { scaleFinder } from "../../Utilities/";

class XAxis extends Component{
    static defaultProps = {
        dimensions:
        {
            width: window.innerWidth*0.9,
            height: window.innerHeight*0.9,
            padding: 50
        },
        className: null,
    }

    static propTypes = {
        data: PropTypes.array.isRequired,
        dimensions: PropTypes.shape({
            width: PropTypes.number,
            height: PropTypes.number,
            padding: PropTypes.number
        }),
        scaleType: PropTypes.oneOf(["linear", "time", "ordinal"]).isRequired,
        className: PropTypes.string,
        aes: PropTypes.string.isRequired
    }


    getXScale(){
        // get props and aesthetic to render ( x-vaue )
        let { data, aes } = this.props;

        // get x-values
        let xValues  = data.map(item => item[aes]);
        let scaleObj = new scaleFinder(xValues);

        return scaleObj;
    }

    getScaleType(){
        const { dimensions, scaleType } = this.props;
        const scaleObj                  = this.getXScale();
        let xScale;

        if(scaleType === "linear")
            xScale = scaleObj.getLinearScale();

        if(scaleType === "time")
            xScale = scaleObj.getTimeScale();

        // needs works
        //if(scaleType === "ordinal")
        //    xScale = scaleObj.getOrdinalScale();

        // set scale range
        xScale.range([dimensions.padding, dimensions.width - dimensions.padding]);

        return xScale;
    }

    findXAxis(){
        let { dimensions } = this.props;
        let xScale         = this.getScaleType();
        let axisLocation   = `translate(0, ${dimensions.height - dimensions.padding})`;

        // select node returned by component and appends x-axis
        select(this.node)
            .attr("transform", axisLocation)
            .call(axisBottom(xScale));
    }

    render(){
        return(
            <g
                ref={ node => this.node = node }
                className={ this.props.className }
            >
            </g>
        );
    }

    componentDidMount(){
        this.findXAxis();
    }

    componentDidUpdate(){
        this.findXAxis();
    }
}

export default XAxis;
