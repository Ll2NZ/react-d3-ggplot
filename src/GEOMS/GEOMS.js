import React            from "react";
import PropTypes        from "prop-types";
import { ChartContext } from "../_context/";

function GEOMS(props) {
        return(
            <ChartContext.Provider value={ props }>
                <svg width={ props.dimensions.width } height={ props.dimensions.height }>
                    { props.children }
                </svg>
            </ChartContext.Provider>
        );
}

GEOMS.propTypes = {
    dimensions: PropTypes.object,
    children: PropTypes.any
};

export default GEOMS;
