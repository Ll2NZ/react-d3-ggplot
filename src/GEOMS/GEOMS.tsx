import React, { ReactChildren, ReactComponentElement, ReactChild } from "react";
import { ChartContext, IContext } from "../_context/";

interface IProps extends IContext {
  children?: ReactChild[] | ReactChild;
}

function GEOMS(props: IProps) {
  if (!props.data || !props.aes || !props.dimensions) {
    throw new Error(`GEOMS must have the following properties: data, aes, and dimension`);
  }

  if (!props.children) {
    throw new Error("GEOMS expects at least one child components such as: <Rects/> or <Points/>");
  }

  return (
    <ChartContext.Provider value={props}>
      <svg width={props.dimensions.width} height={props.dimensions.height}>
        {props.children}
      </svg>
    </ChartContext.Provider>
  );
}

export default GEOMS;
