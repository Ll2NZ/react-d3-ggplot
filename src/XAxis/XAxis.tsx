import React from "react";
import { useContext } from "react";
import { useEffect } from "react";
import { useRef } from "react";
import { ChartContext } from "../_context/";
import { IContext } from "../_context/";
import { useScale } from "../_hooks/";
import { select } from "d3-selection";
import { axisBottom } from "d3-axis";
import { format } from "d3-format";

interface IProps {
  label: string;
}

// componet using application context
function XAxis(props: IProps) {
  // here we use any so we can use ref in select(). if we don't, we can only use a STRING to select a node.
  // And we don't want that since we'll be accidentally touching other <g> elements
  const ref: any = useRef(null);
  const context = useContext<IContext>(ChartContext);
  const scale = useScale(context, "XAxis");

  useEffect(() => {
    const { dimensions, aes, data } = context;
    const axisLocation = `translate(0, ${dimensions.height - dimensions.padding})`;

    // select node returned by component and appends x-axis
    const node = select<SVGGElement, number | Date | string>(ref.current).attr(
      "transform",
      axisLocation
    );

    if (scale !== undefined) {
      node.call(axisBottom(scale));
    }

    // format x-labels
    if (typeof data[0][aes[0]] === "number" && props.label !== undefined) {
      node.selectAll<SVGTextElement, number>("text").html(datum => format(props.label)(datum));
    }
  });

  return <g ref={ref} />;
}

export default XAxis;
