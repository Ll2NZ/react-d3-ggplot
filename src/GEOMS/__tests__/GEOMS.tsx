import React from "react";
import GEOMS from "../GEOMS";
import { IContext } from "../../_context";
import { render } from "react-testing-library";
import { cleanup } from "react-testing-library";
import { XAxis } from "../../XAxis";

afterEach(cleanup);

const notValidState = {
  data: [{ x: "0", y: "0" }, { x: "10", y: "10" }],
  aes: ["x", "y"]
};

// data is expected to have the following shape.
const validState: IContext = {
  data: [{ x: "0", y: "0" }, { x: "10", y: "10" }], // objects in array can have arbitrary property names & multiple properties
  aes: ["x", "y"], // these correspond to the property names above that we want to visualize
  dimensions: { width: 500, height: 300, padding: 50 }
};

test("<GEOMS /> throws an error when an invalid state object is passed in", () => {
  const node = () => {
    render(<GEOMS {...notValidState} />); // ts-lint complains here. however we need to test the invalid state shape
  };

  expect(node).toThrow();
});

test("<GEOMS/> throws an error when no props.children are passed in", () => {
  const node = () => {
    render(<GEOMS {...validState} />);
  };

  expect(node).toThrow();
});

test("<GEOMS/> uses ChartContext.Provider internally", () => {
  const { getByText } = render(
    <GEOMS {...validState}>
      <div>
        this node is used as a place holder so we don't fail. The components passed to GEOMS use the
        context value provided by GEOMS.
      </div>
    </GEOMS>
  );
});
