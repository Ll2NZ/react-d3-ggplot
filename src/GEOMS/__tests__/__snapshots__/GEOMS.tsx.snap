// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GEOMS accepting valid components 1`] = `
<svg
  height="300"
  width="500"
>
  <g
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="middle"
    transform="translate(0, 250)"
  >
    <path
      class="domain"
      d="M50.5,6V0.5H450.5V6"
      stroke="currentColor"
    />
    <g
      class="tick"
      opacity="1"
      transform="translate(50.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(90.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        1
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(130.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        2
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(170.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        3
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(210.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        4
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(250.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        5
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(290.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        6
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(330.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        7
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(370.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        8
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(410.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        9
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(450.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        10
      </text>
    </g>
  </g>
  <g
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="end"
    transform="translate(50, 0)"
  >
    <path
      class="domain"
      d="M-6,250.5H0.5V50.5H-6"
      stroke="currentColor"
    />
    <g
      class="tick"
      opacity="1"
      transform="translate(0,250.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        5.0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,230.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        5.5
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,210.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        6.0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,190.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        6.5
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,170.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        7.0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,150.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        7.5
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,130.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        8.0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,110.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        8.5
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,90.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        9.0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,70.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        9.5
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,50.5)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        10.0
      </text>
    </g>
  </g>
  <g>
    <circle
      cx="50"
      cy="250"
      fill="orange"
      r="3"
    />
    <circle
      cx="450"
      cy="50"
      fill="orange"
      r="3"
    />
  </g>
</svg>
`;
